lib/http_aws_sigv4.c

```diff
@@ -66,6 +66,89 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+/*
+ * Decide in an encoding-independent manner whether a character in an
+ * URL must be escaped. The same criterion must be used in strlen_uri()
+ * and strcpy_uri().
+ */
+static bool urlchar_needs_escaping(int c, bool in_query)
+{
+  return !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
+           (c >= '0' && c <= '9') || c == '-' || c == '.' || c == '_'
+           || c == '~' ||
+           (!in_query && c == '/') ||
+           (in_query && c == '&'));
+}
+
+/*
+ * strlen_uri() returns the length of the given URI
+ * this fucntion is a modified version of strlen_url in urlapi.c
+ */
+static size_t strlen_uri(const char *url, bool in_query)
+{
+  const unsigned char *ptr;
+  size_t newlen = 0;
+  int count_equal = 0;
+
+  for(ptr = (unsigned char *)url; *ptr; ptr++) {
+    switch (*ptr) {
+    case '=':
+      if(!in_query || count_equal++)
+        newlen += 2;
+      newlen++;
+      break;
+    case '&':
+      count_equal = 0;
+      /* FALLTHROUGH */
+    default:
+      if(urlchar_needs_escaping(*ptr, in_query))
+        newlen += 2;
+      newlen++;
+    }
+  }
+  return newlen;
+}
+
+/* strcpy_uri() copies a url to a output buffer and URL-encodes
+ * the URI acordingly
+ * this fucntion is a modified version of strcpy_url in urlapi.c
+ */
+static void strcpy_uri(char *optr, const char *url, bool in_query)
```

Comment:
This function doesn't "copy a URI" it (re-)encodes a query string according to the aws sig v4 rules.

How about instead call it encode_query() ?